be/src/exec/vectorized/sorting/sort_column.cpp

```diff
@@ -97,6 +111,229 @@ class ColumnSorter final : public ColumnVisitorAdapter<ColumnSorter> {
+                containers.emplace_back(&real->get_data());
+            }
+            auto inlined = _create_inlined_permutation<T>(containers);
+            RETURN_IF_ERROR(sort_and_tie_helper(_cancel, &column, _is_asc_order, inlined, _tie, cmp, _range, _build_tie,
+                                                _limit, &_limited));
+            _restore_inlined_permutation(inlined);
+        } else {
+            using ItemType = PermutationItem;
+            auto cmp = [&](const ItemType& lhs, const ItemType& rhs) {
+                auto left_column = down_cast<const ColumnType*>(_vertical_columns[lhs.chunk_index].get());
+                auto right_column = down_cast<const ColumnType*>(_vertical_columns[rhs.chunk_index].get());
+                auto left_value = left_column->get_data()[lhs.index_in_chunk];
+                auto right_value = right_column->get_data()[rhs.index_in_chunk];
+                return SorterComparator<T>::compare(left_value, right_value);
+            };
+
+            RETURN_IF_ERROR(sort_and_tie_helper(_cancel, &column, _is_asc_order, _permutation, _tie, cmp, _range,
+                                                _build_tie, _limit, &_limited));
+        }
+        _prune_limit();
+        return Status::OK();
+    }
+
+    Status do_visit(const vectorized::ConstColumn& column) {
+        // noop
+        return Status::OK();
+    }
+
+    Status do_visit(const vectorized::ArrayColumn& column) {
+        auto cmp = [&](const PermutationItem& lhs, const PermutationItem& rhs) {
+            auto& lhs_col = _vertical_columns[lhs.chunk_index];
+            auto& rhs_col = _vertical_columns[rhs.chunk_index];
+            return lhs_col->compare_at(lhs.index_in_chunk, rhs.index_in_chunk, *rhs_col, _is_null_first ? -1 : 1);
+        };
+
+        RETURN_IF_ERROR(sort_and_tie_helper(_cancel, &column, _is_asc_order, _permutation, _tie, cmp, _range,
+                                            _build_tie, _limit, &_limited));
+        _prune_limit();
+        return Status::OK();
+    }
+
+    template <typename T>
+    Status do_visit(const vectorized::ObjectColumn<T>& column) {
+        DCHECK(false) << "not supported";
+        return Status::NotSupported("TOOD");
+    }
+
+    Status do_visit(const vectorized::JsonColumn& column) {
+        auto cmp = [&](const PermutationItem& lhs, const PermutationItem& rhs) {
+            auto& lhs_col = _vertical_columns[lhs.chunk_index];
+            auto& rhs_col = _vertical_columns[rhs.chunk_index];
+            return lhs_col->compare_at(lhs.index_in_chunk, rhs.index_in_chunk, *rhs_col, _is_null_first ? -1 : 1);
+        };
+
+        RETURN_IF_ERROR(sort_and_tie_helper(_cancel, &column, _is_asc_order, _permutation, _tie, cmp, _range,
+                                            _build_tie, _limit, &_limited));
+        _prune_limit();
+        return Status::OK();
+    }
+
+private:
+    template <class T>
+    struct InlineChunkItem {
```

Comment:
inline has this meaning？ compact ？flatten?

Reply:
There're two types of permutation, one is **out-of-line**, the other is **inline**.

```c++
struct PermtutationItem {
    uint32_t chunk_index; 
    uint32_t index_in_chunk;
};

template <class T>
struct InlinePermutationItem {
    uint32_t index_in_chunk;

    T inline_value;
};
```

Reply:
I know  what you mean. what I mean is the naming issue.

Reply:
Fine, let's rename to `Compact`